upload/upload/main.py

```diff
@@ -0,0 +1,202 @@
+        return self.data['contents']
+
+    def preview(self):
+        s = self.contents()
+        if len(s) > 40:
+            s = s[:40] + " ..."
+        
+        return re.sub('\\s+', ' ', s)
+        
+        return s
+
+app = Flask('upload')
+
+with open('/flask-secret-key/flask-secret-key', 'rb') as f:
+    app.secret_key = f.read()
+
+def get_flow(state=None):
+  flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(
+    CLIENT_SECRET, scopes=SCOPES, state=state)
+
+  # FIXME _scheme='https'
+  flow.redirect_uri = url_for('oauth2callback', _external=True, _scheme='https')
+  
+  return flow
+
+@app.route('/', methods=['GET'])
+def index():
+    if 'email' not in session:
+        return redirect(url_for('login'))
+
+    email = session['email']
+    
+    # just show 50
+    page_items = items[:50]
+    
+    return render_template('index.html', email=email, page_items=page_items, n_items=len(items))
+
+@app.route('/login', methods=['GET'])
+def login():
+  flow = get_flow()
+  authorization_url, state = flow.authorization_url(
+    access_type='offline',
+    include_granted_scopes='true')
+
+  session['state'] = state
+
+  return redirect(authorization_url)
+
+@app.route('/oauth2callback', methods=['GET'])
+def oauth2callback():
+  authorization_response = request.url
+  # FIXME as with url_for(..., _scheme='https')
+  authorization_response = re.sub('^http://', 'https://', authorization_response)
+  
+  flow = get_flow(state=session['state'])
+  flow.fetch_token(authorization_response=authorization_response)
+  credentials = flow.credentials
+  
+  oauth2v2 = build('oauth2', 'v2', credentials=credentials)
+  profile = oauth2v2.userinfo().v2().me().get().execute()
+  email = profile['email']
+  
+  if email in USERS:
+    session['credentials'] = {
+      'token': credentials.token,
+      'refresh_token': credentials.refresh_token,
+      'token_uri': credentials.token_uri,
+      'client_id': credentials.client_id,
+      'client_secret': credentials.client_secret,
+      'scopes': credentials.scopes
+    }
+    session['email'] = email
+    
+    return redirect(url_for('index'))
+  else:
+    abort(401)
+
+@app.route('/upload', methods=['POST'])
+def upload():
+    data = request.json
+    Item(data)
+    return jsonify({})
+
+@app.route('/items/<int:item_id>/contents', methods=['GET'])
+def get_item_contents(item_id):
+    if 'email' not in session:
+        return redirect(url_for('login'))
+
+    item = id_item.get(item_id)
+    if not item:
+        abort(404)
+    return Response(item.contents(), mimetype='text/plain')
+
+@app.route('/items/<int:item_id>/config', methods=['GET'])
+def get_item_config(item_id):
+    if 'email' not in session:
+        return redirect(url_for('login'))
+
+    item = id_item.get(item_id)
+    if not item:
+        abort(404)
+    return Response(json.dumps(item.config(), indent=2), mimetype='text/plain')
+
+@app.route('/logout', methods=['GET'])
+def logout():
+  # FIXME revoke
```

Comment:
as in make their oauth token invalid? I sort of feel like if we don't do it now it will lay here as a FIXME for a while.